sdk/common/src/main/scala/org/typelevel/otel4s/sdk/Attributes.scala

```diff
@@ -75,32 +66,153 @@ final class Attributes private (
+      *   attributes.
+      *
+      * @param attributes
+      *   the attributes to add
+      */
+    def addAll(attributes: Attributes): Builder
+
+    /** Creates [[Attributes]] with the attributes of this builder.
+      */
+    def result(): Attributes
+  }
+
+  val Empty: Attributes = new MapAttributes(Map.empty)
+
+  /** Creates [[Attributes]] with the given `attributes`.
+    *
+    * @note
+    *   if there are duplicated keys in the given `attributes`, only the last
+    *   occurrence will be retained.
+    *
+    * @param attributes
+    *   the attributes to use
+    */
+  def apply(attributes: Attribute[_]*): Attributes =
+    fromIterable(attributes)
+
+  /** Creates [[Attributes]] from the given collection of `attributes`.
+    *
+    * @note
+    *   if there are duplicated keys in the given `attributes`, only the last
+    *   occurrence will be retained.
+    *
+    * @param attributes
+    *   the attributes to use
+    */
+  def fromIterable(attributes: Iterable[Attribute[_]]): Attributes = {
     val builder = Map.newBuilder[AttributeKey[_], Attribute[_]]
     attributes.foreach { a =>
       builder += (a.key -> a)
     }
-    new Attributes(builder.result())
+    new MapAttributes(builder.result())
   }
 
+  /** Creates [[Attributes]] from the given map of `attributes`.
+    *
+    * @param attributes
+    *   the attributes to use
+    */
+  def fromMap(attributes: Map[AttributeKey[_], Attribute[_]]): Attributes =
+    new MapAttributes(attributes)
+
+  /** Creates an empty [[Builder]] of [[Attributes]].
+    */
+  def builder: Builder =
+    new MutableBuilder
+
   implicit val showAttributes: Show[Attributes] = Show.show { attributes =>
     attributes.toList
       .map(a => show"$a")
       .mkString("Attributes(", ", ", ")")
   }
 
   implicit val hashAttributes: Hash[Attributes] =
-    Hash.by(_.m)
+    Hash.by(_.toMap)
 
   implicit val monoidAttributes: Monoid[Attributes] =
     new Monoid[Attributes] {
       def empty: Attributes = Attributes.Empty
-      def combine(x: Attributes, y: Attributes): Attributes = {
+      def combine(x: Attributes, y: Attributes): Attributes =
         if (y.isEmpty) x
         else if (x.isEmpty) y
-        else new Attributes(x.m ++ y.m)
-      }
+        else new MapAttributes(x.toMap ++ y.toMap)
     }
+
+  private final class MapAttributes(
+      private val m: Map[AttributeKey[_], Attribute[_]]
+  ) extends Attributes {
+    def get[T: KeySelect](name: String): Option[Attribute[T]] = {
+      val key = KeySelect[T].make(name)
+      m.get(key).map(_.asInstanceOf[Attribute[T]])
+    }
+
+    def get[T](key: AttributeKey[T]): Option[Attribute[T]] =
+      m.get(key).map(_.asInstanceOf[Attribute[T]])
+
+    def contains(key: AttributeKey[_]): Boolean = m.contains(key)
+
+    def toMap: Map[AttributeKey[_], Attribute[_]] = m
+
+    def iterator: Iterator[Attribute[_]] = m.values.iterator
+
+    override def toList: List[Attribute[_]] = m.values.toList
```

Comment:
The compilation fails without this override in Scala 3:
```
[error] 177 |  private final class MapAttributes(
[error]     |                      ^
[error]     |class MapAttributes needs to be abstract, since def toList: List[org.typelevel.otel4s.Attribute[?]] in trait Attributes in package org.typelevel.otel4s.sdk is not defined 
[error] one error found
```

Reply:
... what?? that makes no sense.

the only thing I can think of is that `Iterable` already defines `toList`, but that doesn't really explain things